portal/portal-service-impl/impl/src/java/org/sakaiproject/portal/service/PortalServiceImpl.java

```diff
@@ -779,14 +780,29 @@ else if (iconType.equalsIgnoreCase("cl")){
 	}
 
 	private boolean canUserUpdateSite(String userId, String siteId) {
-		return securityService.unlock(userId, SiteService.SECURE_UPDATE_SITE, siteService.siteReference(siteId));
+		try {
```

Comment:
this was the performance killer which is why I changed this to a security service check, are you saying this check is unpredictable?

Reply:
right, you're now loading the entire site, pages, tools, properties, members

Reply:
Oddly, securityService.unlock will return 'true' for student members of a site during scheduled publish events or scheduled unpublish events. This is counter-intuitive especially because securityService.unlock returns false (as expected) during the events of publishing and unpublishing a site manually.

Fortunately, for scheduled publish events, canUserUpdateSite is not as significant a determinant like it is for scheduled unpublish events.

I haven't begun adding debugging SakaiSecurity.unlock to assess what's causing that function to return true for student members. It that method is indeed a bug, that would obviously be something to investigate and fix. 

However, if that function should be left as is, then one approach (a kind of hack) would be to distinguish scheduled unpublish events (site.unpublish.scheduled) from their manual analog (site.unpublish) in order execute the more performance costly logic only when responding to scheduled unpublish.

Thoughts?

Reply:
the first thing I think of is the use of a SecurityAdvisor that is set to allow everything .... so during the publishing process it doesn't filter and just says TRUE for everything .... where is the code? my guess is you will find a SecurityAdvisor

Reply:
In the kernel SakaiSecurity.java (approx. line 728) does indeed consult SecurityAdvisor for a few conditions. Empirically it seems to be a factor in returning 'true' for the unlock method during scheduled unpublishing/publishing.

How should I proceed without risking introducing regressions elsewhere?

Reply:
A security advisor is being invoked for the 'execute' method for both classes:

site-manage/site-manage-impl/impl/src/java/org/sakaiproject/sitemanage/impl/PublishingSiteScheduleServiceImpl.java
site-manage/site-manage-impl/impl/src/java/org/sakaiproject/sitemanage/impl/UnpublishingSiteScheduleServiceImpl.java

(I'm omitting for now manual publishing, but that may be a case to return to because it's likely very common for instructors to manually publish their sites and then schedule the unpublishing of their sites-- even if they have to set some arbitrary date in the past for scheduled publishing in order to schedule a date in the future for unpublishing.)

Reply:
@hornersa using a SecurityAdvisor is generally the trivial way to get the access that is needed, I usually attempt to only use a SecurityAdvisor as a last resort. As it very often leads to undesirable side effects like this.

Reply:
I believe two possible fixes:

1) The job should run as admin user like most jobs in jobscheduler (scheduler-component-shared/src/java/org/sakaiproject/component/app/scheduler/jobs/coursepublish/CourseSitePublishJob.java)

2) Modify the SecurityAdvisor check to look for an empty of null userId (e.g., the job). So if the userId isNotBlank then it fails the advisor check. If it is blank, then ALLOWED

Reply:
@ern , @ottenhoff Thanks for the advice, though I think I need to let someone else proceed with implementing a fix. I'll keep watching the jira and will help with QA if available.

Reply:
you're at the two yard line!

    Session session = sessionManager.getCurrentSession();
    try {
      session.setUserEid("admin");
      session.setUserId("admin");
      
      } finally { session.clear(); }

Reply:
@ottenhoff Sorry, I don't follow the additional clue.

Reply:
remove all of the SecurityAdvisor use

then you need to run the job as a user. take a look at a job in the jobscheduler directory. just grep for session.setUserEid

then the job will run as a real user (admin user) instead of running with the advisors

Reply:
I've now force pushed suggested revisions.

Reply:
nice job. i assume all working well in your testing?

Reply:
Thanks, yes. For me this passed the test plan in SAK-49716.